secrets.go

```diff
@@ -0,0 +1,313 @@
+			continue
+		}
+
+		if data, ok := secret.Data["data"]; ok && data != nil {
+			var val interface{}
+			switch data.(type) {
+			case map[string]interface{}:
+				val = data.(map[string]interface{})[keyForVault]
+			}
+
+			secret := fmt.Sprintf("%v", val)
+			if secret != "" {
+				golog.Infof("Found secret for EnvVar: [%s], Key: [%s] in endpoint [%s]", key, keyForVault, endpoint)
+				secrets[keyAsProp] = secret
+			}
+			secrets[keyAsProp] = fmt.Sprintf("%v", val)
+		}
+
+	}
+
+	return secrets, nil
+}
+
+// AzureConfiguration holds azure configuration details
+type AzureConfiguration struct {
+	ClientID     string
+	ClientSecret string
+	TenantID     string
+}
+
+// KeyVault holds the information for a keyvault instance
+type KeyVault struct {
+	client   *keyvault.BaseClient
+	vaultURL string
+}
+
+// ReadLines reads a file and returns the string contents
+func ReadLines(path string) ([]string, error) {
+	file, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	var lines []string
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		lines = append(lines, scanner.Text())
+	}
+	return lines, scanner.Err()
+}
+
+// AzureKeyVaultHandler retrieves secret key values from Azure KeyVault based on environment variables
+func AzureKeyVaultHandler(vaultURL, file string, config AzureConfiguration) (map[string]string, error) {
+	var secretVars []string
+	secrets := make(map[string]string)
+
+	// load secrets from environment or file
+	secretVars, err := loadSecrets(file)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// get a vault client
+	client, err := newKeyVaultClient(vaultURL, config)
+
+	if err != nil {
+		golog.Errorf("Failed to get Azure key client for vault [%s]. [%s]", vaultURL, err.Error())
+		return nil, err
+	}
+
+	for _, v := range secretVars {
+		split := strings.Split(v, "=")
+		key := split[0]
+		keyForAz := strings.ToLower(strings.Replace(key, "_", "-", -1))
+		keyAsProp := strings.ToLower(strings.Replace(key, "_", ".", -1))
+		golog.Infof(fmt.Sprintf("Retrieving secret from vault [%s]. EnvVar: [%s], Key: [%s]", vaultURL, v, keyForAz))
+		secret, err := client.getSecret(keyForAz)
+
+		if err != nil {
+			golog.Errorf("Failed to retrieve secret for key [%s] from vault [%s]. Error [%s]", keyForAz, vaultURL, err.Error())
+			continue
+		}
+
+		if secret != "" {
+			golog.Infof("Found secret for EnvVar: [%s], Key: [%s] in vault [%s]", key, keyForAz, vaultURL)
+			secrets[keyAsProp] = secret
+		}
+	}
+
+	return secrets, nil
+}
+
+// NewKeyVaultClient creates a new keyvault client
+func newKeyVaultClient(vaultURL string, config AzureConfiguration) (*KeyVault, error) {
+
+	keyClient := keyvault.New()
+	credentials := azureAuth.NewClientCredentialsConfig(config.ClientID, config.ClientSecret, config.TenantID)
+	// set the correct resource for keyvault
+	credentials.Resource = "https://vault.azure.net"
```

Comment:
this could be a `const` instead or may even to be configurable via an environment or file. May later will change the resource

Reply:
Done